clippy_lints/src/types/redundant_allocation.rs

```diff
@@ -25,19 +25,40 @@ pub(super) fn check(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>, qpath: &QPath<'_
         }
     }
 
-    if cx.tcx.is_diagnostic_item(sym::Rc, def_id) {
-        if let Some(ty) = is_ty_param_diagnostic_item(cx, qpath, sym::Rc) {
-            let mut applicability = Applicability::MachineApplicable;
-            span_lint_and_sugg(
-                cx,
-                REDUNDANT_ALLOCATION,
-                hir_ty.span,
-                "usage of `Rc<Rc<T>>`",
-                "try",
-                snippet_with_applicability(cx, ty.span, "..", &mut applicability).to_string(),
-                applicability,
-            );
-            true
+    let sym_outer = if cx.tcx.is_diagnostic_item(sym::Rc, def_id) {
+        Some(sym::Rc)
+    } else if cx.tcx.is_diagnostic_item(sym::Arc, def_id) {
+        Some(sym::Arc)
+    } else {
+        None
+    };
+
+    if let Some(sym_outer) = sym_outer {
```

Comment:
I think it would still be better to move the content of this entire if block into a new function for readability. The current function is still quite long, and it's not clear right away that this checks reference counting types.

But that's not totally necessary :)

Reply:
I have made this function cleaner in latest commit.

Reply:
Yes, the changes are really cool, you can resolve this comment. :+1: